proposal/custom-url-resolution.md

```diff
@@ -0,0 +1,117 @@
+# Custom url resolution: Draft 1
+
+_[(Issue)](https://github.com/sass/sass/issues/2535)_
+
+## Table of Contents
+
+- [Background](#background)
+- [Summary](#summary)
+  - [Steps](#steps)
+  - [JavaScript API](#javaScript-api)
+  - [Edge cases](#edge-cases)
+- [Syntax](#syntax)
+- [Deprecation Process](#deprecation-process)
+
+## Background
+
+> This section is non-normative.
+
+Many css features require the use of a url import to reference resources from outside the sass files, however these files also need to exist on the eventual output directory and server. To ensure the references are valid, the sass API should allow for the user to provide a way to remap and/or inline these resources.
+
+## Summary
+
+> This section is non-normative.
+
+This proposal defines a standardized way to remap the url imports to the final location on the server or inline reference.
+
+This is accomplished by running url references through the importer plugin(s) and url rewriting plugin(s) if any url rewriting plugin has been defined.
+
+### Steps
+
+The steps of a url reference import:
+
+- Url reference gets extracted from the Sass file (For more information see [Syntax](#syntax))
+- This reference gets remapped to the actual location on the filesystem by the importer plugin(s)
+- The resolved reference gets passed into the rewrite url plugin pipeline which returns the output url of this resource.
+
+If a certain resource cannot be remapped the url rewrite plugin will return `null`, if all rewrite plugins return `null` it should fallback to the original value of the url reference.
+
+### JavaScript API
+
+At the core of remapping the url imports is the JavaScript API which allows users to return a new url reference based on the filepath or content.
+
+The first parameter this rewriteUrl function gets passed in is the importerResult, which is an object containing an optional `file` and `content` field.
+
+- The `file` field is the location on disk of the given url resource, this can be used to create relative mappings or to simply read the file.
+- The `content` field is a `Blob` which contains the content of the url resource, in this case the url rewriter should not try to read the file manually and use this instead as this is what the importer returned.
```

Comment:
This seems to assume that the Sass implementation already knows how to find a resource on disk given a URL. But as far as Sass is concerned, `url()`s are essentially opaque strings. The goal of a URL rewriter plugin is to encapsulate exactly that knowledge, the knowledge of the relationship between the `url()` string that appears in a file and the actual file that the user wants to refer to.